tasks/migrate-certificates.ts

```diff
@@ -389,106 +423,247 @@ export const GET_MIGRATE_CERTIFICATES_TASK = () =>
           `Signer does not have the CONSIGNOR role in the removal contract`
         );
       }
-      const outputData: MigratedCertificates = [];
-      logger.info(`✨ Migrating ${inputData.length} legacy certificates...`);
+      let outputData: MigratedCertificates = [];
+
+      const alreadyMigrated = (await certificateContract.totalMinted())
+        // eslint-disable-next-line unicorn/no-await-expression-member -- script
+        .toNumber();
+      logger.info(
+        `Found ${alreadyMigrated} certificates already migrated. Minting remaining ${
+          originalInputData.length - alreadyMigrated
+        } certificates...`
+      );
+      if (alreadyMigrated > 0) {
+        let alreadyMintedResults;
+        try {
+          alreadyMintedResults = readJsonSync(outputFile);
+        } catch (error) {
+          logger.error(
+            `${alreadyMigrated} certificates have already been migrated, but no ${outputFile} file found... Something is wrong -- where are the existing results?`
+          );
+          throw error;
+        }
+        if (alreadyMintedResults.length !== alreadyMigrated) {
+          logger.error(
+            `${alreadyMigrated} certificates have already been migrated, but the ${outputFile} file contains ${alreadyMintedResults.length} results... Something is wrong -- why is there a mismatch?`
+          );
+          throw new Error(
+            `${alreadyMigrated} certificates have already been migrated, but the ${outputFile} file contains ${alreadyMintedResults.length} results... Something is wrong -- why is there a mismatch?`
+          );
+        }
+        outputData = outputData.concat(alreadyMintedResults);
+      }
+      const inputData = originalInputData.slice(alreadyMigrated);
+      // const certificateLengths = inputData
+      //   .map((certificate) => certificate.ids.length)
+      //   .sort((a, b) => b - a);
+      // console.log(certificateLengths.slice(0, 10));
+      // return;
+
+      const multicallBatches = [];
+      const batchSize = 100;
+      let batch = [];
+      let singularBatchCount = 0;
+      for (const certificate of inputData) {
+        if (batch.length === batchSize) {
+          multicallBatches.push(batch);
+          batch = [];
+        }
+        if (certificate.ids.length > 30) {
+          if (batch.length > 0) {
+            multicallBatches.push(batch);
+          }
+          batch = [certificate];
+          multicallBatches.push(batch);
+          singularBatchCount += 1;
+          batch = [];
+        }
+        batch.push(certificate);
+      }
+      multicallBatches.push(batch);
```

Comment:
:cold_sweat:

Reply:
In the future, refer to to the [chunk utility in the mono repo](https://github.com/nori-dot-eco/nori/blob/master/nori-graphql/src/utils.ts#L75) for how to chunk arrays

Reply:
yeah that's a clean implementation... this was harder because it's not necessarily creating equally sized chunks - it's dealing with large certificates in a separate way